chainer/functions/caffe/caffe_function.py

```diff
@@ -0,0 +1,336 @@
+import collections
+import sys
+
+import numpy
+
+from chainer import function
+from chainer import function_set
+from chainer import functions
+# caffe_pb2 does not support Py3
+if sys.version_info < (3, 0, 0):
+    from chainer.functions.caffe import caffe_pb2
+
+    _type_to_method = {}
+    _oldname_to_method = {}
+
+    def _layer(typ, oldname):
+        def decorator(meth):
+            global _type_to_method
+            _type_to_method[typ] = meth
+            typevalue = getattr(caffe_pb2.V1LayerParameter, oldname)
+            _oldname_to_method[typevalue] = meth
+            return meth
+        return decorator
+else:
+    def _layer(typ, oldname):  # fallback
+        def decorator(meth):
+            return meth
+        return decorator
+
+
+class CaffeFunction(function.Function):
```

Comment:
This function just inserts a computational graph on top of the given variables in __call__, instead of inserting itself into the graph. Backward computation is done by each node in this graph, so CaffeFunction itself does not need to handle the backward.